[ENH] transformer reconcilers - add tests and improve

ciaran-g opened the pull request:
#### Reference Issues/PRs
<!--
See #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Addresses a few bullet points in #2416

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->

- Removed a restriction on the `Reconciler` which required the bottom level nodes to have unique names by changing the `_get_s_matrix()` function
- Added check for any difference between indexes used in `.fit()` and `.transform()`
- Test the reconciled predictions actually sum appropriately
- Test the index/columns on the summation (S) and G matrices match
- Test the methods work equally for named and unnamed indexes



#### Does your contribution introduce a new dependency? If yes, which one?

<!--
If your contribution does add a new hard dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core sktime package to a minimum.
-->

No

#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

The added tests :-)

fkiraly commented on sktime/transformations/hierarchical/reconcile.py:
```diff
@@ -61,7 +66,7 @@ class Reconciler(BaseTransformer):
     METHOD_LIST = ["bu", "ols", "wls_str"]
 
-    def __init__(self, method="bu"):
+    def __init__(self, method="ols"):
```
we cannot change the default without deprecation now. The reconcilers are being used since 0.11.2, so this change would require a deprecation message, which is displayed whenever `"bu"` is used. Are we certain we want this change in default?

ciaran-g replied:
Good point, I think the `"bu"` is the more intuitive as well - changed back for now. Although `"ols"` may be the preferred method for users in the long-term since it *typically* is more skillful - forecasts at all levels are used in the reconciliation process.

fkiraly commented on sktime/transformations/hierarchical/reconcile.py:
```diff
@@ -94,13 +99,14 @@ def _fit(self, X, y=None):
-        #   if not, add totals to X
-        if not _check_index_good(X):
+        # check index for no "__total", if not add totals to X
+        if _check_index_no_total(X):
```
I don't think this warning is sensible for `_fit`, since nothing is returned. The user would be confused about where things are added.

fkiraly reviewed (changes requested):
Overall, looks good!

Blocking change requests are minor, but blocking as it would break downwards compatibility:

* default parameter change, if desired, must be preceded by a deprecation cycle. Is it sensible to change the default? I think `"bu"`, as current, is a more sensible default because it is very quick? What is your reasoning to change this?

Non-blocking changes:

* the warning in `_fit` seems confusing, I would recommend not to raise it.
* explanation of `s_matrix` return in `_get_s_matrix` got deleted, is that an accident? Every input and output should have an explanation, which `s_matrix` would lose. Further, kindly describe entries, indices, and how it relates to input. Optimally, link to a place in the Hyndman paper to avoid lengthy definitions.
* same for `g_matrix` in `_get_g_matrix_bu`.

ciaran-g commented:
I posted my reasoning for the default change in the comment above @fkiraly, although I didn't think about deprecation. Happy to change back as in the latest commit.

Added more info to the doc strings and removed the warning as well.

fkiraly commented:
you say you changed it back, but the PR still shows `"ols"` as a default - did you forget to push?

fkiraly reviewed (approved):
Thanks, looks good to me now!

fkiraly commented:
(hm, apparently the pre-PR state had docstring default inconsistent with init default - good that this is fixed)
